File: backend-javascript/03-backend-javascript-postgresql/src/modules/profession/profession.controller.js
import { HTTP_STATUS } from '../../shared/constants/http.js';

const MESSAGES = {
  ITEM_NOT_FOUND: 'Profession not found',
};

class Controller {
  constructor(service) {
    this.service = service;

    this.getItems = this.getItems.bind(this);
    this.getItemsCount = this.getItemsCount.bind(this);
    this.getItemsPaginated = this.getItemsPaginated.bind(this);
    this.getItemById = this.getItemById.bind(this);
    this.createItem = this.createItem.bind(this);
    this.updateItem = this.updateItem.bind(this);
    this.deleteItem = this.deleteItem.bind(this);
  }

  async getItems(req, res, next) {
    try {
      console.log('00000000001:' + JSON.stringify(req.query));
      const result = await this.service.getItems(req.query);

      return res.status(HTTP_STATUS.OK).json(result);
    } catch (error) {

      return next(error);
    }
  }

  async getItemsCount(req, res, next) {
    try {
      const result = await this.service.getItemsCount(req.query);

      return res.status(HTTP_STATUS.OK).json(result);
    } catch (error) {

      return next(error);
    }
  }

  async getItemsPaginated(req, res, next) {
    try {
      const result = await this.service.getItemsPaginated(req.query);

      return res.status(HTTP_STATUS.OK).json(result);
    } catch (error) {

      return next(error);
    }
  }

  async getItemById(req, res, next) {
    try {
      const result = await this.service.getItemById(parseInt(req.params.id));
      if (!result) {
        return next({ status: HTTP_STATUS.NOT_FOUND, message: MESSAGES.ITEM_NOT_FOUND });
      }

      return res.status(HTTP_STATUS.OK).json(result);
    } catch (error) {

      return next(error);
    }
  }

  async createItem(req, res, next) {
    try {
      const result = await this.service.createItem(req.body);

      return res.status(HTTP_STATUS.CREATED).json(result);
    } catch (error) {

      return next(error);
    }
  }

  async updateItem(req, res, next) {
    try {
      const result = await this.service.updateItem(parseInt(req.params.id), req.body);
      if (!result) {
        return next({ status: HTTP_STATUS.NOT_FOUND, message: MESSAGES.ITEM_NOT_FOUND });
      }

      return res.status(HTTP_STATUS.OK).json(result);
    } catch (error) {

      return next(error);
    }
  }

  async deleteItem(req, res, next) {
    try {
      const result = await this.service.deleteItem(parseInt(req.params.id));
      if (!result) {
        return next({ status: HTTP_STATUS.NOT_FOUND, message: MESSAGES.ITEM_NOT_FOUND });
      }

      return res.status(HTTP_STATUS.OK).json(result);
    } catch (error) {

      return next(error);
    }
  }
}

export default Controller;

File: backend-java/03-backend-java/src/main/java/com/ganatan/modules/profession/ProfessionController.java
package com.ganatan.modules.profession;

import java.io.IOException;
import java.util.Map;
import java.util.Optional;
import jakarta.servlet.ServletException;
import jakarta.servlet.annotation.WebServlet;
import jakarta.servlet.http.HttpServlet;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.ganatan.config.AppConfig;

@WebServlet("/professions/*")
public class ProfessionController extends HttpServlet {

    private final ProfessionService service;
    private final ObjectMapper mapper = new ObjectMapper();

    public ProfessionController() {
        ProfessionRepository repository = new ProfessionRepository(AppConfig.useDbClient());
        System.out.println("00000000001:ProfessionController");
        this.service = new ProfessionService(repository);
    }

    @Override
    protected void doOptions(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
        setCorsHeaders(response);
        response.setStatus(HttpServletResponse.SC_OK);
    }

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
    	System.out.println("00000000001:ProfessionController:doGet");
        setCorsHeaders(response);
        response.setContentType("application/json");
        response.setCharacterEncoding("UTF-8");

        String pathInfo = request.getPathInfo();

        if (pathInfo == null || pathInfo.equals("/")) {
            int page = parseIntOrDefault(request.getParameter("page"), 1);
            int perPage = parseIntOrDefault(request.getParameter("perPage"), 5);
            int offset = (page - 1) * perPage;

            Map<String, Object> result = service.getItems(offset, perPage);
            mapper.writeValue(response.getWriter(), result);
        } else {
            try {
                int id = extractId(pathInfo);
                Optional<Profession> item = service.getItemById(id);
                if (item.isPresent()) {
                    mapper.writeValue(response.getWriter(), item.get());
                } else {
                    response.setStatus(HttpServletResponse.SC_NOT_FOUND);
                    writeError(response, "Item not found");
                }
            } catch (NumberFormatException e) {
                response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
                writeError(response, "Invalid ID");
            }
        }
    }

    @Override
    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
        setCorsHeaders(response);
        Profession profession = mapper.readValue(request.getReader(), Profession.class);
        Profession created = service.createItem(profession);

        response.setStatus(HttpServletResponse.SC_CREATED);
        response.setContentType("application/json");
        mapper.writeValue(response.getWriter(), created);
    }

    @Override
    protected void doPut(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
        setCorsHeaders(response);
        String pathInfo = request.getPathInfo();

        try {
            int id = extractId(pathInfo);
            Profession updatedData = mapper.readValue(request.getReader(), Profession.class);
            Optional<Profession> updated = service.updateItem(id, updatedData);

            if (updated.isPresent()) {
                response.setContentType("application/json");
                mapper.writeValue(response.getWriter(), updated.get());
            } else {
                response.setStatus(HttpServletResponse.SC_NOT_FOUND);
                writeError(response, "Item not found");
            }
        } catch (NumberFormatException e) {
            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
            writeError(response, "Invalid ID");
        }
    }

    @Override
    protected void doDelete(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
        setCorsHeaders(response);
        String pathInfo = request.getPathInfo();

        try {
            int id = extractId(pathInfo);
            boolean deleted = service.deleteItem(id);

            if (deleted) {
                response.setStatus(HttpServletResponse.SC_OK);
                mapper.writeValue(response.getWriter(), mapper.createObjectNode().put("message", "Deleted successfully"));
            } else {
                response.setStatus(HttpServletResponse.SC_NOT_FOUND);
                writeError(response, "Item not found");
            }
        } catch (NumberFormatException e) {
            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
            writeError(response, "Invalid ID");
        }
    }

    private int extractId(String pathInfo) {
        String[] parts = pathInfo.split("/");
        return Integer.parseInt(parts[1]);
    }

    private int parseIntOrDefault(String value, int defaultValue) {
        try {
            return Integer.parseInt(value);
        } catch (NumberFormatException | NullPointerException e) {
            return defaultValue;
        }
    }

    private void writeError(HttpServletResponse response, String message) throws IOException {
        ObjectNode error = mapper.createObjectNode();
        error.put("error", message);
        mapper.writeValue(response.getWriter(), error);
    }

    private void setCorsHeaders(HttpServletResponse response) {
        response.setHeader("Access-Control-Allow-Origin", "http://localhost:4200");
        response.setHeader("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS");
        response.setHeader("Access-Control-Allow-Headers", "Content-Type, Authorization");
        response.setHeader("Access-Control-Allow-Credentials", "true");
    }
}
